doc/settings.md
# CHEM Settings

Use **Settings** to control Eagan Matrix settings not available in the other CHEM modules.

![CHEM Settings module](./image/settings.png)

Click the **fingerboard** to reverse the direction of the playing surface.

For **Middle C**, Turn the knob or right click to enter the MIDI note number (nn) that has the frequency of Middle C. In other words, transpose the surface.

**Touch Area** is the MIDI Note number for the center of the touch area, from 16 to 108. A number <= 16 means there is no touch area. Turn the knob or right click to enter the MIDI note number.

Next select the **X** (bend range), and the **Y** (front-back) and **Z** (pressure) MIDI cc assiignments. The last section for **X** is a combination with MPE (96) for MPE channels 2+, and a selected range on Channel 1.

**Note processing** lets you determine how notes are processed, including the MIDI *Note On* velocity:

| Option | Description |
| -- | -- |
| **Static velocity** | All notes have a static velocity of 127, the maximum. |
| **Dynamic velocity** | Calculated from the initial attack on the fingerboard. |
| **Formula velocity** | Calculated using EM formula V. |
| **No notes** | Supresses all production of MIDI notes. |
| **Ethervox** | All note information is a pitch bend from MIDI Note 60 (middle C). |
| **Kyma** | Special mode for controlling a Symbolic Sound Kyma device. |

The **Mono** button turns Mono mode on and off.

The **Mono mode** menu offers a variety of modes for how a note transitions from one to another when Mono mode is on.

**Mono Interval** controls the interval from the first note where the mono behavior takes effect.
The default is off, where Mono applies across the entire playing surface when Mono is on.

The **Tuning** menu selects the tuning for notes from the playing surface.
Choose from *n*-TET (equal) tunings,
Just intonation for selected scales,
or custom tunings created using the 3rd-party *Pitch Table Editor* (PTE).
See the Haken Audio documentation for more information on using the PTE.

The **Rounding** menu offers choices for when round is applied.

The **Round Initial** check box controls the obvious thing.

The **Round rate** slider controls the rounding rate from 0 (none) to 10 (fastest).
The selected rounding rate can be modulated using the **RATE** input.

The **Keep MIDI** check box is the equivalent of *Preserve MIDI* in the Haken Editor's *Global Settings*.

The **Keep Surface** check box is the equivalent of *Preserve Surface* in the Haken Editor.

The equivalent to the Haken Editor's *Preserve Pedals* setting is on the [**Jack** module](./jack.md).

The **MIDI routing** diagram is above the inputs sections. Click a light to disable/enable routing from the input on the left (*MIDI* or **Surface*) to the output on the bottom (*DSP*, *CVC*, and *MIDI*).

The AES menu provides access to AES audio input settings, when available for a device.

## Inputs

The **INIT** input accepts a unipolar 0 to 10v gate to control the **Round Initial** setting.
When the gate is high, Round Initial is on.
When the gate is low, Round Initial is off.

The **RATE** input is a bipolar (-5 to +5v) modulation of the selected **Round rate**.

---

// Copyright © Paul Chase Dempsey\
![pachde (#d) logo](./image/Logo.svg)


src/modules/Settings/Settings.cpp
#include "Settings.hpp"
#include "../../services/rack-help.hpp"
#include "../../em/wrap-HakenMidi.hpp"
#include "tuning.hpp"
using namespace pachde;

SettingsModule::SettingsModule() :
    modulation(this, ChemId::Settings)
{
    config(Params::NUM_PARAMS, Inputs::NUM_INPUTS, Outputs::NUM_OUTPUTS, Lights::NUM_LIGHTS);

    std::vector<std::string> offon{"off", "on"};
    std::vector<std::string> onoff{"on", "off"};

    configSwitch(P_SURFACE_DIRECTION, 0.f, 1.f, 0.f, "Surface direction", {"normal", "reversed"});
    // TODO: intervals continuous from 1 to 96 in engine?
    configSwitch(P_X, 0.f, 15.f, 11.f, "X Bend Range", {
        "1", "2", "3", "4", "5", "6", "7", "12", "24", "36", "48", "96",
        "96: 2 (MPE)",
        "96: 5",
        "96: 7",
        "96: 12"
    });
    configSwitch(P_Y, 0.f, 8.f, 7.f, "Y assign", {"off", "cc 1 (mod wheel)", "cc 2 (breath)", "cc 3", "cc 4 (foot)", "cc 7 (vol)", "cc 11 (expression)", "cc 74 (MPE*)", "cc 74 (no shelf)"});
    configSwitch(P_Z, 0.f, 9.f, 8.f, "Z assign", {"off", "cc 1 (mod wheel)", "cc 2 (breath)", "cc 3", "cc 4 (foot)", "cc 7 (vol)", "cc 11 (expression)", "Channel pressure", "MPE+ (14-bit)", "MPE (Channel pressure)"});
    configSwitch(P_NOTE_PROCESSING, 0.f, 5.f, 0.f, "Note processing", {
        "Static velocity (127)", 
        "Dynamic velocity",
        "Formula velocity (ƒV)",
        "No Notes",
        "Ethervox",
        "Kyma"
    });
    configSwitch(P_NOTE_PRIORITY, 0.f, 6.f, 0.f, "Note Priority", {
        "Oldest channel",
        "Same pitch",
        "Lowest channel",
        "Highest channel",
        "Highest 2 channels",
        "Highest 3 channels",
        "Highest 4 channels",
    });
    configSwitch(P_BASE_POLYPHONY, 0.f, 7.f, 3.f, "Base polyphony", {"1", "2", "3", "4", "5", "6", "7", "8"});
    configSwitch(P_EXPAND_POLYPHONY, 0.f, 1.f, 0.f, "Expand polyphony", offon);
    configSwitch(P_DOUBLE_COMPUTATION, 0.f, 1.f, 0.f, "Double computation rate", offon);
    configSwitch(P_MONO, 0.f, 1.f, 0.f, "Mono", offon);
    configSwitch(P_MONO_MODE, 0.f, 8.f, 0.f, "Mono mode", {
        "Weighted portamento",
        "Legato to high pressure",
        "Retrigger high pressure",
        "Legato to touch",
        "Retrig touch",
        "Retrig touch & lift",
        "Legato to touch, max Z",
        "Retrig touch, max Z",
        "Retrig touch & lift, max Z"
    });
    configSwitch(P_MONO_INTERVAL, 0.f, 14.f, 0.f, "Mono interval", {
        "Off",
        "1 semitone",
        "Second",
        "Minor third",
        "Major third",
        "Fourth",
        "Augmented fourth",
        "Fifth",
        "Sixth",
        "Minor seventh",
        "Major seventh",
        "1 octave",
        "2 octaves",
        "3 octaves",
        "Maximum",
    });
    configSwitch(P_OCTAVE_SWITCH, 0.f, 1.f, 0.f, "Octave switch", offon);
    configSwitch(P_OCTAVE_TYPE, 0.f, 2.f, 0.f, "Octave switch type", {
        "Switch",
        "Toggle",
        "Instant"
    });
    configSwitch(P_OCTAVE_RANGE, 0.f, 7.f, 3.f, "Octave switch range", {
        "-4 octaves",
        "-3 octaves",
        "-2 octaves",
        "-1 octave",
        "+1 octave",
        "+2 octaves",
        "+3 octaves",
        "+4 octaves",
    });
    configSwitch(P_ROUND_TYPE, 0.f, 3.f, 0.f, "Round type", {
        "Normal",
        "Release",
        "Y",
        "Reverse Y",
    });
    configSwitch(P_ROUND_INITIAL, 0.f, 1.f, 0.f, "Round initial", offon);
    configParam(P_ROUND_RATE, 0.f, 10.f, 0.f, "Round rate");

    configTuningParam(this, P_TUNING);

    configSwitch(P_MIDI_DSP,     0.f, 1.f, 1.f, "Route Midi to DSP",     offon);
    configSwitch(P_MIDI_CVC,     0.f, 1.f, 1.f, "Route Midi to CVC",     offon);
    configSwitch(P_MIDI_MIDI,    0.f, 1.f, 1.f, "Route Midi to MIDI",    offon);
    configSwitch(P_SURFACE_DSP,  0.f, 1.f, 1.f, "Route Surface to DSP",  offon);
    configSwitch(P_SURFACE_CVC,  0.f, 1.f, 1.f, "Route Surface to CVC",  offon);
    configSwitch(P_SURFACE_MIDI, 0.f, 1.f, 1.f, "Route Surface to MIDI", offon);
    configSwitch(P_KEEP_MIDI,    0.f, 1.f, 0.f, "Keep MIDI settings",    offon);
    dp4(configParam(P_MOD_AMOUNT, -100.f, 100.f, 0.f, "Modulation amount", "%")); 


    configInput(IN_ROUND_INITIAL, "Round initial gate");
    configInput(IN_ROUND_RATE, "Round rate");

    // button lights unconfigured to disable tooltip
    //configLight(L_MONO, "Mono mode");
    //configLight(L_OCTAVE_SWITCH, "Octave switch");

    configLight(L_ROUND_Y,       "Round on Y");
    configLight(L_ROUND_INITIAL, "Round initial");
    configLight(L_ROUND,         "Rounding");
    configLight(L_ROUND_RELEASE, "Round on release");

    // EmccPortConfig cfg[] = {
    // };
    //modulation.configure(Params::P_MOD_AMOUNT, NUM_MOD_PARAMS, cfg);
}

bool SettingsModule::connected()
{
    if (!chem_host) return false;
    auto conn = chem_host->host_connection(ChemDevice::Haken);
    return conn && conn->identified();
}

void SettingsModule::dataFromJson(json_t *root)
{
    ChemModule::dataFromJson(root);
    json_t* j = json_object_get(root, "haken-device");
    if (j) {
        device_claim = json_string_value(j);
    }
    if (!device_claim.empty()) {
        modulation.mod_from_json(root);
    }

    ModuleBroker::get()->try_bind_client(this);
}

json_t* SettingsModule::dataToJson()
{
    json_t* root = ChemModule::dataToJson();
    json_object_set_new(root, "haken-device", json_string(device_claim.c_str()));
    if (!device_claim.empty()) {
        modulation.mod_to_json(root);
    }
    return root;
}

void SettingsModule::update_from_em()
{
    if (!connected()) return;
    auto em = chem_host->host_matrix();
    if (!em->is_ready()) { return; }

    //modulation.set_em_and_param_low(P_PRE_LEVEL, em->get_pre(), true);
}

void SettingsModule::do_message(PackedMidiMessage message)
{
    if (Haken::ccStat1 != message.bytes.status_byte) return;
    if (as_u8(ChemId::Settings) == message.bytes.tag) return;

//    modulation.set_em_and_param_low(param, midi_cc_value(message), true);
}

// IChemClient
::rack::engine::Module* SettingsModule::client_module() { return this; }
std::string SettingsModule::client_claim() { return device_claim; }

void SettingsModule::onConnectHost(IChemHost* host)
{
    onConnectHostModuleImpl(this, host);
}

void SettingsModule::onPresetChange()
{
    if (connected()) {
        update_from_em();
        //if (chem_ui) ui()->onPresetChange(); // ui does nothing
    }
}

void SettingsModule::onConnectionChange(ChemDevice device, std::shared_ptr<MidiDeviceConnection> connection)
{
    if (chem_ui) ui()->onConnectionChange(device, connection);
}

void SettingsModule::process_params(const ProcessArgs &args)
{
    // auto v = getParam(Params::P_MIX).getValue()*.1f;
    // getLight(Lights::L_MIX).setBrightnessSmooth(v, 45.f);

    //modulation.pull_mod_amount();
}

void SettingsModule::process(const ProcessArgs &args)
{
    ChemModule::process(args);

    if (!connected() || chem_host->host_busy()) return;

    if (0 == ((args.frame + id) % 45)) {
        process_params(args);
    }

    // if (modulation.sync_params_ready(args)) {
    //      modulation.sync_send();
    // }

    if (((args.frame + id) % 61) == 0) {
        //modulation.update_lights();

        auto em = chem_host->host_matrix();
        float rate   = em->get_round_rate() / 127.f;
        bool initial = em->is_round_initial();
        bool on_y    = em->get_round_mode() >= Haken::rViaY;
        bool release = em->get_round_mode() <= Haken::rRelease;
        getLight(Lights::L_ROUND_Y).setBrightness(1.0f * on_y);
        getLight(Lights::L_ROUND_INITIAL).setBrightness(1.0f * initial);
        getLight(Lights::L_ROUND).setBrightness(1.0f * rate);
        getLight(Lights::L_ROUND_RELEASE).setBrightness(1.0f * ((rate > 0.f) && release));
        
        float v = getParam(P_MONO).getValue();
        getLight(Lights::L_MONO).setBrightness(v);

        v = getParam(P_OCTAVE_SWITCH).getValue();
        getLight(Lights::L_OCTAVE_SWITCH).setBrightness(v);

        for (int p = Params::P_MIDI_DSP; p <= Params::P_SURFACE_MIDI; ++p) {
            v = getParam(p).getValue();
            getLight(p - Params::P_MIDI_DSP + L_MIDI_DSP).setBrightness(v);
        }
    }

}

Model *modelSettings = createModel<SettingsModule, SettingsUi>("chem-settings");

